### ISSUE TITLE
Frequency-based trips

### ISSUE THREAD

**Issue by irees** [transitland/transitland-datastore#408]
Handle frequency.txt style trips, e.g. 
https://rawgit.com/CUTR-at-USF/bullrunner-gtfs-realtime-generator/master/bullrunner-gtfs.zip

**Comment by irees** [comment]
Investigate multiplying-trips vs. internally supporting frequency.txt based trips.

**Comment by barbeau** [comment]
I would suggest taking a look at OpenTripPlanner, as they've dealt with this same question.  I haven't delved into the code myself, but a student from our group did some prototyping in OpenTripPlanner for supporting frequency-based `exact_times=0` real-time info (https://github.com/opentripplanner/OpenTripPlanner/pull/1647), and also created a frequency-based `exact_times=0` GTFS-realtime feed for the USF Bull Runner (https://github.com/CUTR-at-USF/bullrunner-gtfs-realtime-generator).

**Comment by drewda** [comment]
@irees @gknisely while this ticket waits on the backburner, likely until next quarter, let's downgrade all frequency-based feeds to import level 2, so they are not imported into Valhalla for routing.

**Comment by irees** [comment]
11 currently active feeds that use `frequencies.txt` -- trip_count is the total number of trips.txt trips, frequencies_count is the number of frequencies.txt entries.

```[{
	"onestop_id": "f-drt-mbta",
	"active_feed_version": "9c1b5facfed0b7cab2742492124f42989d487b7b",
	"trip_count": 158417,
	"frequencies_count": 21
}, {
	"onestop_id": "f-f25d-socitdetransportdemontral",
	"active_feed_version": "b3cb8fd3828ef9dcabc07b4bd5b86cdbd3478f02",
	"trip_count": 127566,
	"frequencies_count": 1032
}, {
	"onestop_id": "f-9qh0-anaheim~ca~us",
	"active_feed_version": "c93d84a5208adc5d4a4e6729a10a5d35c24683d0",
	"trip_count": 13616,
	"frequencies_count": 1
}, {
	"onestop_id": "f-s-anm~it",
	"active_feed_version": "0977871b4b1767955025d9e90853c4326b71146a",
	"trip_count": 4756,
	"frequencies_count": 1573
}, {
	"onestop_id": "f-9g3-semovi",
	"active_feed_version": "fd847780ff1c19fdc45756fa47a489a5c0d75b01",
	"trip_count": 1679,
	"frequencies_count": 1679
}, {
	"onestop_id": "f-kzf-nairobi~saccos",
	"active_feed_version": "fdb6aad5e0c21c010414eed2e7245786330911ad",
	"trip_count": 268,
	"frequencies_count": 804
}, {
	"onestop_id": "f-9w28-naipta~az~us",
	"active_feed_version": "0e475297f9e8842754f447f74ec256fa9fa68520",
	"trip_count": 125,
	"frequencies_count": 126
}, {
	"onestop_id": "f-dnjj-clemson~sc~us",
	"active_feed_version": "db1eac59d735f378629c47a18722f0bff56a7425",
	"trip_count": 94,
	"frequencies_count": 45
}, {
	"onestop_id": "f-69y7-recursosdatabuenosairesgobar",
	"active_feed_version": "95742a1a65826abfa23f5aeaeeb082e45f066801",
	"trip_count": 48,
	"frequencies_count": 172
}, {
	"onestop_id": "f-9xhv-nps~romo~shuttles",
	"active_feed_version": "55b23bcff5ccf13a08e1b1fbe49308e34f086ebd",
	"trip_count": 16,
	"frequencies_count": 16
}, {
	"onestop_id": "f-u3qc-metrowarszawskie",
	"active_feed_version": "3a7d41d38edf4e1d1d3340103cc6ab5245e3128b",
	"trip_count": 16,
	"frequencies_count": 98
}]```

**Comment by barbeau** [comment]
Are you all currently supporting frequency-based `exact_times=1` trips (i.e., the compressed representation of schedule-based trips)?  Or are these on back burner for now as well?

**Comment by irees** [comment]
My current concept is to describe frequency-based trips using "ScheduleStopPair" attributes "frequency_headway_seconds", "frequency_start_time", and "frequency_end_time", with arrival & departure times relative to the start of the trip. I don't currently have a flag for `exact_times` but it could be added.

For example, an SSP with frequency_headway_seconds=600, frequency_start_time=08:00:00, and origin_departure_time=00:05:00 would result in departures at 08:05, 08:15, 08:25, etc.

We've been discussing this lately, and I welcome any feedback and suggestions.

**Comment by barbeau** [comment]
@irees I'm not very familiar with the internal data model of transitland, but what you describe seems to make sense for feeds with `exact_times=1`.  This is effectively a compressed way to represent a fixed schedule without including a bunch of near duplicate trips, and most transit software I've seen treats it exactly as you describe.

However, `exact_times=0` (or if `exact_times` is omitted) is a different beast.  This is true frequency-based, meaning that the expected arrival time of the vehicle shifts throughout the day, based on when the vehicle actually arrives.

This has a lot of implications when you start getting into user-facing information, such as answering the question "when is my bus expected to arrive?"  For `exact_times=1`, the answer is the same as normal scheduled trips.  For example, assuming lack of real-time info, if it's 8:20, "your bus is scheduled to arrive in five minutes".  But for `exact_times=1`, the answer is "your bus is expected to arrive every 600 seconds."  And, if you don't have real-time info, that answer is the same no matter when during the day you ask the question.  For routing and calculating wait times, without real-time info systems like OTP use a pessimistic estimate and assume the wait is always the headway.  If you have real-time info with a predicted arrival time, then you materialize an estimated arrival time on the fly based on this estimate - in that case, then you can say "your bus is expected to arrive in five minutes".

This seems to be a subtle difference, but in creating a GTFS-rt frequency-based exact_times=0 feed, and trying to add support for this into OTP, we ran into a lot of unexpected issues.  IMHO, true frequency-based systems are tricky.

**Comment by drewda** [comment]
@irees the one final step that can happen on another issue: adding mention of the new frequency attributes to the SSP docs: https://github.com/transitland/www-transit-land/issues/274

**Comment by irees** [comment]
Frequency-based schedules are now supported in the datastore & in routing.

Example:
https://valhalla.mapzen.com/route?json={%22locations%22:[{%22lat%22:28.05732490546724,%22lon%22:-82.40939140319824,%22type%22:%22break%22,%22name%22:%22%22},{%22lat%22:28.080386371114074,%22lon%22:-82.41333961486816,%22type%22:%22break%22,%22name%22:%22%22}],%22costing%22:%22multimodal%22,%22costing_options%22:{%22transit%22:{%22use_bus%22:%221.0%22,%22use_rail%22:%220.0%22,%22use_transfers%22:%221.0%22}},%22directions_options%22:{%22language%22:%22en-US%22},%22date_time%22:{%22type%22:0}}
(add your api key)

![screen shot 2017-03-09 at 4 56 51 pm](https://cloud.githubusercontent.com/assets/720087/23867980/009e03a6-07db-11e7-99fe-8c66220bbd8c.png)

**Comment by barbeau** [comment]
@irees Nice!  So what assumption does Valhalla make for the estimated wait time for the bus for a frequency-based `exact_times` = 0 system like our USF Bull Runner?  IIRC OTP uses a pessimistic approach and assumes you'd need to wait the entire headway defined in GTFS.
